File: tools/parsing-question/src/parser/answer_extractor.py
"""
Answer Extractor

Extracts answers and correctAnswer from LaTeX content based on question type.
"""

import re
from typing import List, Optional, Union, Tuple
import sys
import os

# Add parent directory to path for imports
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)

from models.question import QuestionAnswer
from parser.bracket_parser import BracketParser
from utils.text_cleaner import TextCleaner


class AnswerExtractor:
    """
    Extracts answers and correctAnswer from LaTeX content.
    
    Handles different question types:
    - MC: answers array + correctAnswer string
    - TF: answers array + correctAnswer string array
    - SA: answers null + correctAnswer string
    - ES: answers null + correctAnswer null
    """
    
    # Patterns for question type detection
    CHOICE_PATTERNS = [
        r'\\choice(?:\[[0-9]\])?',
        r'\\choice(?:\s|$)'
    ]
    
    CHOICE_TF_PATTERNS = [
        r'\\choiceTF(?:\[[t12]\])?',
        r'\\choiceTFt',
        r'\\choiceTF(?:\s|$)'
    ]
    
    SHORT_ANS_PATTERNS = [
        r'\\shortans(?:\[[a-z0-9]+\])?',
        r'\\shortans(?:\s|$)'
    ]
    
    MATCHING_PATTERNS = [
        r'\\matching(?:\s|$)',
        r'\\matching(?:\{.*?\})?'
    ]
    
    TRUE_PATTERN = r'\\True'
    SOLUTION_PATTERN = r'\\loigiai\s*\{([^}]*(?:\{[^}]*\}[^}]*)*)\}'
    
    @classmethod
    def identify_question_type(cls, content: str) -> str:
        """
        Identify question type based on LaTeX commands.

        Logic phân loại:
        1. Có "\\choiceTF" → TF (Đúng/Sai)
        2. Có "\\choice" (không phải choiceTF) → MC (Nhiều lựa chọn)
        3. Có "\\shortans" → SA (Trả lời ngắn)
        4. Có "\\matching" → MA (Ghép đôi) - tạm thời skip
        5. Không có gì → ES (Tự luận)

        Args:
            content: LaTeX content to analyze

        Returns:
            Question type: MC, TF, SA, ES
        """
        # Remove solution section for analysis
        content_without_solution = cls._remove_solution_for_analysis(content)

        # 1. Check for choiceTF variants → TF (Đúng/Sai)
        # Includes: \choiceTF, \choiceTFt, \choiceTF[...], etc.
        if '\\choiceTF' in content_without_solution:
            return "TF"

        # 2. Check for choice (not choiceTF) → MC (Nhiều lựa chọn)
        if '\\choice' in content_without_solution:
            return "MC"

        # 3. Check for shortans → SA (Trả lời ngắn)
        if '\\shortans' in content_without_solution:
            return "SA"

        # 4. Check for matching → MA (Ghép đôi) - skip for now
        if '\\matching' in content_without_solution:
            return "ES"  # Treat as essay for now

        # 5. Default to essay
        return "ES"
    
    @classmethod
    def extract_mc_answers(cls, content: str) -> Tuple[List[QuestionAnswer], Optional[str]]:
        """
        Extract answers for Multiple Choice questions.
        
        Args:
            content: LaTeX content
            
        Returns:
            Tuple of (answers_list, correct_answer_content)
        """
        answers = []
        correct_answer = None
        
        # Find \choice command (but not \choiceTF)
        choice_pos = -1
        pos = 0
        while pos < len(content):
            pos = content.find('\\choice', pos)
            if pos == -1:
                break
            # Check if it's not choiceTF
            if not content[pos:pos+9] == '\\choiceTF':
                choice_pos = pos
                break
            pos += 1

        if choice_pos == -1:
            return [], None

        # Find the end of the choice command (after optional parameters)
        start_pos = choice_pos
        while start_pos < len(content) and content[start_pos] not in ['\n', '{']:
            start_pos += 1

        answer_index = 0
        
        while start_pos < len(content):
            # Skip whitespace
            while start_pos < len(content) and content[start_pos].isspace():
                start_pos += 1
            
            if start_pos >= len(content) or content[start_pos] != '{':
                break
            
            # Extract answer content
            answer_content = BracketParser.extract_content_from_braces(content, start_pos)

            # Trim leading/trailing whitespace for checking
            trimmed_content = answer_content.strip()

            # Check if this is the correct answer
            is_correct = False
            if trimmed_content.startswith('\\True '):
                is_correct = True
                answer_content = trimmed_content[6:].strip()  # Remove '\True ' (6 chars)
                # Convert newlines to literal for correct answer
                correct_answer = TextCleaner.convert_newlines_to_literal(answer_content)
            elif trimmed_content.startswith('\\True'):
                is_correct = True
                answer_content = trimmed_content[5:].strip()  # Remove '\True' (5 chars)
                # Convert newlines to literal for correct answer
                correct_answer = TextCleaner.convert_newlines_to_literal(answer_content)
            else:
                # Keep original content but trimmed
                answer_content = trimmed_content

            # Convert newlines to literal for CSV export
            answer_content = TextCleaner.convert_newlines_to_literal(answer_content)

            answers.append(QuestionAnswer(
                id=answer_index,
                content=answer_content,
                isCorrect=is_correct
            ))
            
            # Move past this answer
            start_pos = cls._find_closing_brace(content, start_pos) + 1
            answer_index += 1
        
        return answers, correct_answer
    
    @classmethod
    def extract_tf_answers(cls, content: str) -> Tuple[List[QuestionAnswer], Optional[List[str]]]:
        """
        Extract answers for True/False questions.
        
        Args:
            content: LaTeX content
            
        Returns:
            Tuple of (answers_list, correct_answers_list)
        """
        answers = []
        correct_answers = []
        
        # Find \choiceTF command
        choice_tf_pos = content.find('\\choiceTF')
        if choice_tf_pos == -1:
            return [], None

        # Find the end of the choiceTF command (after optional parameters)
        start_pos = choice_tf_pos
        while start_pos < len(content) and content[start_pos] not in ['\n', '{']:
            start_pos += 1
        
        # Extract answers after \choiceTF
        answer_index = 0
        
        while start_pos < len(content):
            # Skip whitespace
            while start_pos < len(content) and content[start_pos].isspace():
                start_pos += 1
            
            if start_pos >= len(content) or content[start_pos] != '{':
                break
            
            # Extract answer content
            answer_content = BracketParser.extract_content_from_braces(content, start_pos)

            # Trim leading/trailing whitespace for checking
            trimmed_content = answer_content.strip()

            # Check if this is a correct answer
            is_correct = False
            if trimmed_content.startswith('\\True '):
                is_correct = True
                answer_content = trimmed_content[6:].strip()  # Remove '\True ' (6 chars)
                # Convert newlines to literal for correct answers
                correct_answer_literal = TextCleaner.convert_newlines_to_literal(answer_content)
                correct_answers.append(correct_answer_literal)
            elif trimmed_content.startswith('\\True'):
                is_correct = True
                answer_content = trimmed_content[5:].strip()  # Remove '\True' (5 chars)
                # Convert newlines to literal for correct answers
                correct_answer_literal = TextCleaner.convert_newlines_to_literal(answer_content)
                correct_answers.append(correct_answer_literal)
            else:
                # Keep original content but trimmed
                answer_content = trimmed_content

            # Convert newlines to literal for CSV export
            answer_content = TextCleaner.convert_newlines_to_literal(answer_content)
            
            answers.append(QuestionAnswer(
                id=answer_index,
                content=answer_content,
                isCorrect=is_correct
            ))
            
            # Move past this answer
            start_pos = cls._find_closing_brace(content, start_pos) + 1
            answer_index += 1
        
        return answers, correct_answers if correct_answers else None
    
    @classmethod
    def extract_sa_answer(cls, content: str) -> Optional[str]:
        """
        Extract answer for Short Answer questions.

        Format: \\shortans{answer} hoặc \\shortans[param]{answer}

        Args:
            content: LaTeX content

        Returns:
            Correct answer string or None
        """
        # Find \shortans command
        shortans_pos = content.find('\\shortans')
        if shortans_pos == -1:
            return None

        # Start from after \shortans
        pos = shortans_pos + len('\\shortans')

        # Skip optional parameters [...]
        while pos < len(content) and content[pos].isspace():
            pos += 1

        if pos < len(content) and content[pos] == '[':
            # Skip optional parameter [...]
            bracket_level = 1
            pos += 1
            while pos < len(content) and bracket_level > 0:
                if content[pos] == '[':
                    bracket_level += 1
                elif content[pos] == ']':
                    bracket_level -= 1
                pos += 1

        # Skip whitespace again
        while pos < len(content) and content[pos].isspace():
            pos += 1

        # Extract answer from braces {...}
        if pos < len(content) and content[pos] == '{':
            answer = BracketParser.extract_content_from_braces(content, pos)
            if answer:
                # Clean up the answer
                answer = answer.strip()
                # Remove quotes if present
                answer = answer.strip('\'"')
                # Convert newlines to literal for CSV export
                answer = TextCleaner.convert_newlines_to_literal(answer)
                return answer

        return None
    
    @classmethod
    def extract_solution(cls, content: str) -> Optional[str]:
        """
        Extract solution from \\loigiai{...} command.

        Args:
            content: LaTeX content

        Returns:
            Solution content with newlines converted to literal \n
        """
        loigiai_start = content.find('\\loigiai')
        if loigiai_start != -1:
            # Find the opening brace after \loigiai
            brace_start = content.find('{', loigiai_start)
            if brace_start != -1:
                solution = BracketParser.extract_content_from_braces(content, brace_start)
                solution = solution.strip()
                # Convert newlines to literal for CSV export
                solution = TextCleaner.convert_newlines_to_literal(solution)
                return solution

        return None
    
    @classmethod
    def extract_answers_and_correct(cls, content: str, question_type: str) -> Tuple[Optional[List[QuestionAnswer]], Optional[Union[str, List[str]]]]:
        """
        Extract answers and correctAnswer based on question type.
        
        Args:
            content: LaTeX content
            question_type: Question type (MC, TF, SA, ES)
            
        Returns:
            Tuple of (answers, correctAnswer)
        """
        if question_type == "MC":
            answers, correct_answer = cls.extract_mc_answers(content)
            return answers, correct_answer
        
        elif question_type == "TF":
            answers, correct_answers = cls.extract_tf_answers(content)
            return answers, correct_answers
        
        elif question_type == "SA":
            correct_answer = cls.extract_sa_answer(content)
            if correct_answer:
                # Create a single answer object for SA questions
                answer = QuestionAnswer(
                    id=0,
                    content=correct_answer,
                    isCorrect=True
                )
                return [answer], correct_answer
            return [], None
        
        elif question_type == "ES":
            return None, None
        
        else:
            return None, None
    
    @classmethod
    def _remove_solution_for_analysis(cls, content: str) -> str:
        """Remove solution section for type analysis."""
        match = re.search(cls.SOLUTION_PATTERN, content)
        if match:
            return content[:match.start()]
        return content
    
    @classmethod
    def _find_closing_brace(cls, content: str, start_pos: int) -> int:
        """Find the position of the closing brace."""
        if start_pos >= len(content) or content[start_pos] != '{':
            return start_pos
        
        bracket_level = 1
        pos = start_pos + 1
        
        while pos < len(content) and bracket_level > 0:
            if content[pos] == '\\' and pos + 1 < len(content):
                pos += 2
                continue
            if content[pos] == '{':
                bracket_level += 1
            elif content[pos] == '}':
                bracket_level -= 1
            pos += 1
        
        return pos - 1  # Position of closing brace

File: apps/backend/internal/latex/answer_extractor.go
package latex

import (
	"encoding/json"
	"regexp"
	"strings"

	"exam-bank-system/apps/backend/internal/entity"
	"github.com/jackc/pgtype"
)

// AnswerExtractor extracts answers and correct answers from LaTeX content
type AnswerExtractor struct {
	bp *BracketParser
}

// QuestionAnswer represents a single answer option
type QuestionAnswer struct {
	ID        int    `json:"id"`
	Content   string `json:"content"`
	IsCorrect bool   `json:"isCorrect"`
}

// NewAnswerExtractor creates a new answer extractor
func NewAnswerExtractor() *AnswerExtractor {
	return &AnswerExtractor{
		bp: NewBracketParser(),
	}
}

// IdentifyQuestionType identifies question type based on LaTeX commands
func (ae *AnswerExtractor) IdentifyQuestionType(content string) string {
	// Remove solution section for analysis
	contentWithoutSolution := ae.removeSolutionForAnalysis(content)

	// Priority order:
	// 1. Check for choiceTF variants â†’ TF
	if strings.Contains(contentWithoutSolution, "\\choiceTF") {
		return string(entity.QuestionTypeTF)
	}

	// 2. Check for choice (not choiceTF) â†’ MC
	if strings.Contains(contentWithoutSolution, "\\choice") {
		return string(entity.QuestionTypeMC)
	}

	// 3. Check for shortans â†’ SA
	if strings.Contains(contentWithoutSolution, "\\shortans") {
		return string(entity.QuestionTypeSA)
	}

	// 4. Check for matching â†’ MA
	if strings.Contains(contentWithoutSolution, "\\matching") {
		return string(entity.QuestionTypeMA)
	}

	// 5. Default to essay
	return string(entity.QuestionTypeES)
}

// ExtractAnswersAndCorrect extracts answers and correct answer based on question type
func (ae *AnswerExtractor) ExtractAnswersAndCorrect(content, questionType string) (pgtype.JSONB, pgtype.JSONB) {
	var answersJSON, correctAnswerJSON pgtype.JSONB

	switch questionType {
	case string(entity.QuestionTypeMC):
		answers, correctAnswer := ae.extractMCAnswers(content)
		if len(answers) > 0 {
			answersData, _ := json.Marshal(answers)
			answersJSON.Set(answersData)
		}
		if correctAnswer != "" {
			correctAnswerJSON.Set([]byte(`"` + correctAnswer + `"`))
		}

	case string(entity.QuestionTypeTF):
		answers, correctAnswers := ae.extractTFAnswers(content)
		if len(answers) > 0 {
			answersData, _ := json.Marshal(answers)
			answersJSON.Set(answersData)
		}
		if len(correctAnswers) > 0 {
			correctAnswersData, _ := json.Marshal(correctAnswers)
			correctAnswerJSON.Set(correctAnswersData)
		}

	case string(entity.QuestionTypeSA):
		correctAnswer := ae.extractSAAnswer(content)
		if correctAnswer != "" {
			// For SA, create a single answer object
			answers := []QuestionAnswer{
				{ID: 0, Content: correctAnswer, IsCorrect: true},
			}
			answersData, _ := json.Marshal(answers)
			answersJSON.Set(answersData)
			correctAnswerJSON.Set([]byte(`"` + correctAnswer + `"`))
		}

	case string(entity.QuestionTypeES):
		// ES questions don't have predefined answers
		answersJSON.Status = pgtype.Null
		correctAnswerJSON.Status = pgtype.Null
	}

	return answersJSON, correctAnswerJSON
}

// ExtractSolution extracts solution from \loigiai{...} command
func (ae *AnswerExtractor) ExtractSolution(content string) string {
	loigiaiStart := strings.Index(content, "\\loigiai")
	if loigiaiStart == -1 {
		return ""
	}

	// Find the opening brace after \loigiai
	braceStart := strings.Index(content[loigiaiStart:], "{")
	if braceStart == -1 {
		return ""
	}
	braceStart += loigiaiStart

	// Extract solution using bracket parser
	solution := ae.bp.ExtractContentFromBraces(content, braceStart)
	return strings.TrimSpace(solution)
}

// extractMCAnswers extracts answers for Multiple Choice questions
func (ae *AnswerExtractor) extractMCAnswers(content string) ([]QuestionAnswer, string) {
	var answers []QuestionAnswer
	var correctAnswer string

	// Find \choice command (but not \choiceTF)
	choicePos := ae.findChoicePosition(content)
	if choicePos == -1 {
		return answers, correctAnswer
	}

	// Find the end of the choice command (after optional parameters)
	startPos := choicePos
	for startPos < len(content) && content[startPos] != '\n' && content[startPos] != '{' {
		startPos++
	}

	answerIndex := 0

	// Extract answers after \choice
	for startPos < len(content) {
		// Skip whitespace
		for startPos < len(content) && (content[startPos] == ' ' || content[startPos] == '\t' || content[startPos] == '\n') {
			startPos++
		}

		if startPos >= len(content) || content[startPos] != '{' {
			break
		}

		// Extract answer content
		answerContent := ae.bp.ExtractContentFromBraces(content, startPos)
		if answerContent == "" {
			break
		}

		// Trim and check for correct answer
		trimmedContent := strings.TrimSpace(answerContent)
		isCorrect := false

		// Check for \True at the end (most common case)
		if strings.HasSuffix(trimmedContent, " \\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-6]) // Remove ' \True'
			correctAnswer = answerContent
		} else if strings.HasSuffix(trimmedContent, "\\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-5]) // Remove '\True'
			correctAnswer = answerContent
		} else if strings.HasPrefix(trimmedContent, "\\True ") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[6:]) // Remove '\True '
			correctAnswer = answerContent
		} else if strings.HasPrefix(trimmedContent, "\\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[5:]) // Remove '\True'
			correctAnswer = answerContent
		} else {
			answerContent = trimmedContent
		}

		answers = append(answers, QuestionAnswer{
			ID:        answerIndex,
			Content:   answerContent,
			IsCorrect: isCorrect,
		})

		// Move past this answer
		startPos = ae.findClosingBrace(content, startPos) + 1
		answerIndex++
	}

	return answers, correctAnswer
}

// extractTFAnswers extracts answers for True/False questions
func (ae *AnswerExtractor) extractTFAnswers(content string) ([]QuestionAnswer, []string) {
	var answers []QuestionAnswer
	var correctAnswers []string

	// Find \choiceTF command
	choiceTFPos := strings.Index(content, "\\choiceTF")
	if choiceTFPos == -1 {
		return answers, correctAnswers
	}

	// Find the end of the choiceTF command (after optional parameters)
	startPos := choiceTFPos
	for startPos < len(content) && content[startPos] != '\n' && content[startPos] != '{' {
		startPos++
	}

	answerIndex := 0

	// Extract answers after \choiceTF
	for startPos < len(content) {
		// Skip whitespace
		for startPos < len(content) && (content[startPos] == ' ' || content[startPos] == '\t' || content[startPos] == '\n') {
			startPos++
		}

		if startPos >= len(content) || content[startPos] != '{' {
			break
		}

		// Extract answer content
		answerContent := ae.bp.ExtractContentFromBraces(content, startPos)
		if answerContent == "" {
			break
		}

		// Trim and check for correct answer
		trimmedContent := strings.TrimSpace(answerContent)
		isCorrect := false

		// Check for \True at the end (most common case)
		if strings.HasSuffix(trimmedContent, " \\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-6]) // Remove ' \True'
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasSuffix(trimmedContent, "\\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-5]) // Remove '\True'
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasSuffix(trimmedContent, " \\False") {
			// \False is also a correct answer marker for TF questions
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-7]) // Remove ' \False'
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasSuffix(trimmedContent, "\\False") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[:len(trimmedContent)-6]) // Remove '\False'
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasPrefix(trimmedContent, "\\True ") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[6:]) // Remove '\True '
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasPrefix(trimmedContent, "\\True") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[5:]) // Remove '\True'
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasPrefix(trimmedContent, "\\False ") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[7:]) // Remove '\False '
			correctAnswers = append(correctAnswers, answerContent)
		} else if strings.HasPrefix(trimmedContent, "\\False") {
			isCorrect = true
			answerContent = strings.TrimSpace(trimmedContent[6:]) // Remove '\False'
			correctAnswers = append(correctAnswers, answerContent)
		} else {
			answerContent = trimmedContent
		}

		answers = append(answers, QuestionAnswer{
			ID:        answerIndex,
			Content:   answerContent,
			IsCorrect: isCorrect,
		})

		// Move past this answer
		startPos = ae.findClosingBrace(content, startPos) + 1
		answerIndex++
	}

	return answers, correctAnswers
}

// extractSAAnswer extracts answer for Short Answer questions
func (ae *AnswerExtractor) extractSAAnswer(content string) string {
	// Find \shortans command
	shortansPos := strings.Index(content, "\\shortans")
	if shortansPos == -1 {
		return ""
	}

	// Start from after \shortans
	pos := shortansPos + len("\\shortans")

	// Skip optional parameters [...]
	for pos < len(content) && (content[pos] == ' ' || content[pos] == '\t') {
		pos++
	}

	if pos < len(content) && content[pos] == '[' {
		// Skip optional parameter [...]
		bracketLevel := 1
		pos++
		for pos < len(content) && bracketLevel > 0 {
			if content[pos] == '[' {
				bracketLevel++
			} else if content[pos] == ']' {
				bracketLevel--
			}
			pos++
		}
	}

	// Skip whitespace again
	for pos < len(content) && (content[pos] == ' ' || content[pos] == '\t') {
		pos++
	}

	// Extract answer from braces {...}
	if pos < len(content) && content[pos] == '{' {
		answer := ae.bp.ExtractContentFromBraces(content, pos)
		if answer != "" {
			// Clean up the answer
			answer = strings.TrimSpace(answer)
			// Remove quotes if present
			answer = strings.Trim(answer, `'"`)
			return answer
		}
	}

	return ""
}

// findChoicePosition finds \choice command position (but not \choiceTF)
func (ae *AnswerExtractor) findChoicePosition(content string) int {
	pos := 0
	for pos < len(content) {
		choicePos := strings.Index(content[pos:], "\\choice")
		if choicePos == -1 {
			break
		}
		choicePos += pos

		// Check if it's not \choiceTF
		if choicePos+9 >= len(content) || content[choicePos:choicePos+9] != "\\choiceTF" {
			return choicePos
		}
		pos = choicePos + 1
	}
	return -1
}

// removeSolutionForAnalysis removes solution section for type analysis
func (ae *AnswerExtractor) removeSolutionForAnalysis(content string) string {
	re := regexp.MustCompile(`\\loigiai\s*\{.*?\}`)
	matches := re.FindStringSubmatch(content)
	if len(matches) > 0 {
		return content[:strings.Index(content, matches[0])]
	}
	return content
}

// findClosingBrace finds the position of the closing brace
func (ae *AnswerExtractor) findClosingBrace(content string, startPos int) int {
	if startPos >= len(content) || content[startPos] != '{' {
		return startPos
	}

	braceLevel := 1
	pos := startPos + 1

	for pos < len(content) && braceLevel > 0 {
		char := content[pos]

		// Handle escaped characters
		if char == '\\' && pos+1 < len(content) {
			pos += 2
			continue
		}

		if char == '{' {
			braceLevel++
		} else if char == '}' {
			braceLevel--
		}

		pos++
	}

	return pos - 1 // Position of closing brace
}
